Monika After Story/game/zz_apikeys.rpy

```diff
@@ -0,0 +1,556 @@
+                key = key[:Data.MAX_KEY_SIZE_DISP - 3] + "..."
+
+            feats.append((
+                Data.registered_api_keys[feature][0],
+                feature,
+                key,
+            ))
+
+        return sorted(feats)
+
+
+    def has_features():
+        """
+        Do we have API-key based features?
+
+        RETURNS: True if we have api key based features
+        """
+        return len(Data.registered_api_keys) > 0
+
+
+    def clean_key(dirty_key):
+        """
+        Cleans an aPI key
+
+        IN:
+            dirty_key - key to clean
+
+        RETURNS: cleaned key
+        """
+        return dirty_key.replace("\n", "").replace("\r", "")
+
+
+    def _run_on_change(feature, api_key):
+        """
+        Runs on change for a feature with api key
+
+        IN:
+            feature - feature to run on change for
+            api_key - api key to run on change with
+
+        RETURNS: tuple of the following format:
+            [0] - True if valid key, False if not
+            [1] - error message to show
+        """
+        on_change = Data.registered_api_keys[feature][1]
+        if on_change is not None:
+            try:
+                rv = on_change(api_key)
+
+                if not isinstance(rv, tuple) or len(rv) < 2:
+                    return True, ""
+
+                return rv
+
+            except Exception as e:
+                mas_utils.mas_log.error(
+                    "crash when running on_change for feature {0} - {1}".format(
+                        feature,
+                        repr(e)
+                    )
+                )
+
+                return False, "on-change crash - see logs"
+
+        return True, ""
+
+
+    def screen_clear(feature):
+        """
+        Called when the clear button is used on the api key screen
+
+        IN:
+            feature - the feature whose key is being cleared
+        """
+        if not feature_registered(feature) or not store.mas_hasAPIKey(feature):
+            return
+
+        # clear key
+        Data.api_keys.pop(feature)
+        save_keys()
+
+        # on change
+        _run_on_change(feature, "")
+
+
+    def screen_paste(feature):
+        """
+        Called when the paste button is used on the api key screen.
+
+        IN:
+            feature - the feature whose key is being pasted
+        """
+        if not feature_registered(feature):
+            return
+
+        # grab key 
+        new_key = pygame.scrap.get(pygame.SCRAP_TEXT).strip()
+        if not new_key:
+            # null key is not counted
+            return
+
+        # clear newlines
+        new_key = clean_key(new_key)
+
+        # on change
+        key_valid, err_msg = _run_on_change(feature, new_key)
+
+        if key_valid:
+            # set key
+            Data.api_keys[feature] = new_key
+            save_keys()
+
+        else:
+            # show message box
+            store.renpy.show_screen(
```

Comment:
>we don't show dialogues from screens

the settings menu shows a screen for the unstable popup (the dialog box is a screen). the screen being shown here is a similar message popup.

> executing some random code from a screen

I'm confused what you mean by this, the code here is being executed because a screen button was clicked, its just abstracted into a function so it doesnt clutter the screen code like the suntimes does in the settings menu. 

if you mean `on_change` is unsafe, well that's likely to be a needed feature for anyone wanting to use api keys. And a submod trying to use that maliciously could do much worse elsewhere.